### ISSUE TITLE
Dynamics rate control [feature]

### ISSUE THREAD

**Issue by PabloBarddal** [byuu/bsnes#227]
Hi Byuu, First thanks for your work, im really glad to have such accurate emulador for SNES :)

About the issue, Dynamics rate Control is a good way to avoid tearing and audio skips, but as of now only OSS audio is supported under Linux.

I would like to ask about the possibility of supporting dynamic rate control for other audio drivers in Linux, Like alsa or pulse audio (alsa even have a compatibility layer for oss).

I ask this because oss is very old and Very difficult to install in a Lot of distros, If the goal of emulators is preserve the system for the future, we need this feature for more modern audio drivers (also i tested a Lot of configs and i believe the dynamic rate Control must be the only way to get tear free experience in linux… while being a vendor neutral feature).

Thanks for the attention.

**Comment by Screwtapello** [comment]
byuu has implemented DRC for all the platforms he personally has access to - the various audio APIs on Windows, and OSS on FreeBSD. I've asked him about DRC for the PulseAudio driver before, and he's happy to add it if somebody else provides it, but he's not interested in doing the work himself.

**Comment by obunden** [comment]
I don't know if this is something that needs to be implemented for each core. But retroarch should handle this if you don't mind using it instead. If you want a more up to date version consider building the bsnes libretro core yourself. The libretro port available is outdated as far as I know.

**Comment by byuu** [comment]
Basically I would love to have universal DRC. But if you want DRC for another audio API, then I need help implementing it.

XAudio2 and DirectSound: the functions to poll the current playback position do not appear to be precise enough to allow it to work.

WASAPI and ASIO: these drivers are barely usable (the latter I won't even build anymore), so I haven't really investigated it.

ALSA, PulseAudio, PulseAudio-Simple, libao, etc: I don't run Linux.

OpenAL: would be nice for macOS support, but I don't know much about OpenAL.

**Comment by PabloBarddal** [comment]
Openal It would be a good idea, since we have openal (software mode) in Linux, and Works really good for me, even better than the pulse, and much better than alsa.
If windows support openal too we have here the universal solution for drc.

**Comment by obunden** [comment]
Wouldn't just focusing on the actual core and letting frontends like retroarch worry about the other stuff be easier and more efficient? Since it already supports all platforms / sound APIs that I can think off and G-sync/Freesync and Drc? Don't want to step on anyone's toes or anything, there is probably something I'm missing.

**Comment by byuu** [comment]
I already have a libretro port, if people want to use that they already can.

**Comment by PabloBarddal** [comment]
I dont Care about retroarch, i only use bsnes. As of know bsnes by itself cant be tear Free on Linux.

**Comment by obunden** [comment]
Okay fair enough.  Just thought it was worth mentioning since it checked off all the things that you wanted right now until it eventually gets implemented in the regular version. I'm all for giving people the option to choose and I really appreciate that there is a gui version and a libretro core. And most of all that after all this time bsnes/higan still is actively maintained. 

Thanks @byuu and all contributors!

**Comment by mudlord** [comment]
>Basically I would love to have universal DRC. But if you want DRC for another audio API, then I need help implementing it. XAudio2 and DirectSound: the functions to poll the current playback position do not appear to be precise enough to allow it to work.

In my libretro loader, I make a ringbuffer for miniaudio, which is a single-file-header audio library with support for WASAPI/DirectSound/Xaudio2/WinMM/ALSA/OSS/whatever. This ringbuffer has then its buffer status polled as needed, to work out the resampling ratio. This works out fine for my loader since then I don't have to worry about the underlying driver since all I poll is the ring buffer write/read status. So then, DRC works with whatever underlying driver I use for miniaudio.

https://github.com/mudlord/einweggerat/blob/master/io/audio.cpp

Maybe something similar could be done?

**Comment by byuu** [comment]
I was under the impression we needed the actual hardware buffer playback position in order to detect how quickly we are depleting the samples we feed to the sound card, though.

Nonetheless this sounds promising, if it can work I'd happily take it, even if it results in a touch more audio latency, it'd be well worth it.

**Comment by RedDwarf69** [comment]
I hadn't seen this issue. But I have been using DRC with alsa: https://github.com/byuu/bsnes/pull/286

I actually first looked at doing it with PulseAudio. It's not impossible, but as far as I saw
- There are two buffers, the PulseAudio server one and the hardware ring buffer. And PulseAudio doesn't tell you about the state of the hardware one. (but you can guess how big it is)
- You can't even ask PulseAudio how filled its own buffer is. Since PulseAudio request data from you, though, you can deduce how filled the buffer is depending on how much data it has requested. The granularity of these request can be configured, but by default it's quite high... and reducing it obviously will increase the CPU usage since there are more messages being transmitted between the PulseAudio server and client.

By default, with my hardware, if I ask for 20ms latency to PA I get a 15ms PA buffer... so I guess there is a 5ms hardware buffer. And (by default) PA will only request data in groups of 5ms.

So the DRC would need to take the decision based on whether there had been 0, 1, 2 or 3 data requests.

If there are 0 requests it means the PA buffer is more than 10ms filled, and supposedly the hardware buffer fully filled. So there are between 15 and 20 ms of data in the buffers.

If there is 1 request it means the PA buffer is more than 5ms filled, and supposedly the hardware buffer fully filled. So there are between 10 and 15 ms of data in the buffers.

If there are 2 request it means the PA buffer is more than 0ms filled, and supposedly the hardware buffer fully filled. So there are between 5 and 10 ms of data in the buffers.

If there are 3 request it means the PA buffer is empty, and you have no idea of the hardware buffer level. So there are between 0 and 5 ms of data in the buffers.

It's not like you need more than that. The only thing you need to know is whether the buffer is more or less filled than the point you want to achieve (i.e. 50%). But probably a bit more granularity would be good (e.g. more gradual pitch changes? Couldn't say, I'm tone deaf).


From a very quick look at RetroArch it seems they basically do  mudlord's `ringbuffer` thing with the PA buffer. They try to keep the PA buffer half full, ignoring the hardware buffer. Which in the previous example would mean they actually try to keep 12.5 ms instead of 10ms of data in buffers.
They don't try to increase the granularity (https://github.com/libretro/RetroArch/blob/142503d3b54fbb0098244ad2e934ceca3d20d798/audio/drivers/pulse.c#L206) and simply look at the PA requests (https://github.com/libretro/RetroArch/blob/142503d3b54fbb0098244ad2e934ceca3d20d798/audio/drivers/pulse.c#L342). Has anybody tried it? What latency does it use? How well does it work?

Again, doable... but at that point I though better to just use alsa directly. Maybe it's not pro-audio work, but bsnes 20ms request is latency sensitive enough to justify skipping PulseAudio (specially given that the emulator core basically eats all my laptop CPU cores :-p).

**Comment by RedDwarf69** [comment]
Since I already had it I have just created https://github.com/byuu/bsnes/pull/287 also for PulseAudio. It's equivalent to the RetroArch implementation, no clever tricks are attempted.

**Comment by byuu** [comment]
Very much appreciated, thank you! I will try to merge your patches when I have some downtime.

**Comment by RedDwarf69** [comment]
To be honest I ended up building the RetroArch engine and using it instead of the bsnes native interface. But to have some context from what I did before I forget:

- I always had some problems when using DRC, but attributed them to a lack of CPU issue because of https://github.com/byuu/bsnes/issues/289
- Once understood my laptop has **plenty** of spare CPU I looked at it further and noticed https://github.com/byuu/bsnes/pull/286#issuecomment-576060953. Since the video sync is broken (at least in my computer, on Linux, with Wayland) it doesn't matter what it's done in the audio-side, DRC simply doesn't work. (I just though about Wayland, I should try with Xorg)

If the video sync issue is specific of my computer then adding the DRC support to the audio driver may help some people. But RetroArch can make the video sync part work on my computer, so there is something that can be improved in bsnes to make it work for at least another set of people.

I never really tried to look at the video sync issue. I more or less understand audio, but I have no idea where to start to look with the video sync issue. And that's when I gave up and went to RetroArch (I originally found its interface confusing, but once I understood it I was happy enough).
